### PR TITLE
Fix for FuelSDK Bugs

### PR DESCRIPTION
**Pull request by shiva-cirus**
The PR contains fixes for FuelSDK Pagination support .  Bug with ClickEvent ( Unable to retrieve due to naming issues ) and support for NotSent Tracking event that's currently not part of FuelSDK.

### CODE REVIEW

**Review by rmstar — COMMENTED** [review]
Some initial comments, I'll take another look later:
- Copyright © 2020 Cask Data, Inc. -> Copyright © 2021 Cask Data, Inc.
- remove commented out code
- clean up comments
   * space after //
   * remove empty comment lines, etc

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+import com.exacttarget.fuelsdk.internal.Soap;
+import org.apache.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.exacttarget.fuelsdk.ETDataExtension.retrieveColumns;
+
+/**
+ * Extending class to fix the bug with Fuel-SDK on pagination.
```
Comment (rmstar) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:43]: what's the pagination bug how was it fixed? can you add some details in the comemnts

File: src/main/java/com/custom/fuelsdk/ETNotSentEvent.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+import java.util.Date;
+
+/**
+ * An <code>ETNotSentEvent</code> object represents information pertaining to the notsent events
+ * of an email message in the Salesforce Marketing Cloud.
+ * This object type is currently not supported in FuelSDK and hence the provided here.
+ */
+@SoapObject(internalType = NotSentEvent.class)
+public class ETNotSentEvent extends ETSoapObject {
```
Comment (rmstar) [inline src/main/java/com/custom/fuelsdk/ETNotSentEvent.java:33]: where is this used?

File: src/main/java/com/custom/fuelsdk/ETClickEvent.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  private Integer batchID;
+
+  @ExternalName("urlid")
+  private Integer urlid;
+
+  @ExternalName("url")
+  private String url;
+
+  /**
+   * @return the id
```
Comment (rmstar) [inline src/main/java/com/custom/fuelsdk/ETClickEvent.java:63]: I don't think the javadoc is really useful for simple getters and setters.

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+   * @throws ETSdkException
+   */
+  public static ETResponse<ETDataExtensionRow> select(ETClient client,
+                                                      String dataExtension,
+                                                      ETFilter filter)
+    throws ETSdkException {
+    String name = null;
+
+    //
+    // The data extension can be specified using key or name:
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:83]: Could both key and name be specified?

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+      if (filter.getProperties().isEmpty()) {
+        throw new ETSdkException("columns must be specified "
+                                   + "when retrieving data extensions by name");
+      }
+
+    } else {
+      throw new ETSdkException("invalid data extension filter string");
+    }
+
+    ETResponse<ETDataExtensionRow> response = customRetrieve(client, "DataExtensionObject[" + name + "]",
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:155]: any validation on 'name'? Like, what is its value itself is null? Can that happen?

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+                                                      String dataExtension,
+                                                      ETFilter filter)
+    throws ETSdkException {
+    String name = null;
+
+    //
+    // The data extension can be specified using key or name:
+    //
+
+    ETExpression e = ETExpression.parse(dataExtension);
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:86]: Can you add comments here as well, pls.
Like if key is specified then filter can be empty and all columns will be retreived
if name is specified then filter needs to be specified as well.
Also, if there is a api link to point to, that would be great.

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+   * @throws ETSdkException
+   */
+  public static ETResponse<ETDataExtensionRow> continueRequest(ETClient client, String soapObject, String requestID,
+                                                               ETFilter filter) throws ETSdkException {
+    ETResponse<ETDataExtensionRow> response = customRetrieve(client, soapObject, filter, requestID,
+                                                             ETDataExtensionRow.class);
+    return response;
+  }
+
+  //there was a bug with original method related with continueRequest
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java]: Can you add a pointer to what bug? Id or reference, pls

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+
+    //
+    // Read internal type from the SoapObject annotation:
+    //
+
+    Class<T> externalType = type; // for code readability
+
+    SoapObject internalTypeAnnotation
+      = externalType.getAnnotation(SoapObject.class);
+    assert internalTypeAnnotation != null;
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:212]: Why assert? Why not a proper check for the null value?
Same for below.

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+   */
+  public static ETResponse<ETDataExtensionRow> continueRequest(ETClient client, String soapObject, String requestID,
+                                                               ETFilter filter) throws ETSdkException {
+    ETResponse<ETDataExtensionRow> response = customRetrieve(client, soapObject, filter, requestID,
+                                                             ETDataExtensionRow.class);
+    return response;
+  }
+
+  //there was a bug with original method related with continueRequest
+  public static <T extends ETSoapObject> ETResponse<T> customRetrieve(ETClient client,
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:184]: Should this be private?

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+          getInternalProperty(externalType, externalProperty);
+        assert internalProperty != null;
+        internalProperties.add(internalProperty);
+      }
+    } else {
+      //
+      // No properties were explicitly requested:
+      //
+
+      internalProperties = getInternalProperties(externalType);
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:243]: Can this be null or empty?

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+      // Only request those properties specified:
+      //
+
+      internalProperties = new ArrayList<String>();
+
+      for (String externalProperty : externalProperties) {
+        String internalProperty =
+          getInternalProperty(externalType, externalProperty);
+        assert internalProperty != null;
+        internalProperties.add(internalProperty);
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:236]: None of these could be unretreivable, right?
Similar to the check below?

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+
+        retrieveRequest.setFilter(toFilterPart(expression));
+      }
+    } else {
+      if (continueRequest != null) {
+        retrieveRequest.setContinueRequest(continueRequest);
+      }
+    }
+
+//    //todo reduce batchsize
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java]: Pls remove commented code.

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+//    RetrieveOptions retrieveOptions = new RetrieveOptions();
+////    retrieveOptions.setBatchSize(1);
+//    retrieveRequest.setOptions(retrieveOptions);
+
+    if (logger.isTraceEnabled()) {
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:299]: Can you refactor the tracing into a separate method, pls

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+      logger.trace("  objectType = " + retrieveRequest.getObjectType());
+      StringBuilder line = new StringBuilder("  properties = { ");
+      //String line = null;
+      for (String property : retrieveRequest.getProperties()) {
+
+        line.append(property).append(", ");
+      }
+      if (retrieveRequest.getProperties().size() > 0) {
+        line.setLength(line.length() - 2);
+                /*if (line == null) {
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java]: remove dead code, pls

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+        logger.trace("  filter = " + toFilterPart(expression));
+      }
+    }
+
+    logger.trace("calling soap.retrieve...");
+
+    RetrieveRequestMsg retrieveRequestMsg = new RetrieveRequestMsg();
+    retrieveRequestMsg.setRetrieveRequest(retrieveRequest);
+
+    //todo temp fix for continue request
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java]: why a temp fix? Can you add a comment on whats the permanent fix, pls

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+    if (soap == null) {
+      if (soapObjectName != null) {
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:325]: a similar pattern is repeated above, would be good to refactor

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+//        retrieveRequest.setObjectType(soapObjectName);
+        soap = connection.getSoap("retrieve", soapObjectName);
+      } else {
+//        retrieveRequest.setObjectType(internalType.getSimpleName());
+        soap = connection.getSoap("retrieve", internalType.getSimpleName());
+      }
+    }
+    RetrieveResponseMsg retrieveResponseMsg = soap.retrieve(retrieveRequestMsg);
+
+    if (logger.isTraceEnabled()) {
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:334]: refactor to a separate method, pls

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+      logger.trace("  overallStatus = " + retrieveResponseMsg.getOverallStatus());
+      logger.trace("  results = {");
+      for (APIObject result : retrieveResponseMsg.getResults()) {
+        logger.trace("    " + result);
+      }
+      logger.trace("  }");
+    }
+
+    response.setRequestId(retrieveResponseMsg.getRequestID());
+    if (retrieveResponseMsg.getOverallStatus().equals("OK")) {
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:346]: can you make this case insensitive check, pls
same for below.

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+      //
+      // Add result to the list of results:
+      //
+
+      ETResult<T> result = new ETResult<T>();
+      result.setObject(externalObject);
+      response.addResult(result);
+    }
+
+    if (retrieveResponseMsg.getOverallStatus().equals("MoreDataAvailable")) {
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:383]: case insensitive check, pls

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+    return response;
+  }
+
+  /**
+   * @param client The ETClient object
+   * @param key    The key
+   * @return The List of Column names
+   * @throws ETSdkException
+   */
+  private static List<String> retrieveColumnNames(ETClient client,
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:396]: do we really need this method? The below method is enough, no?

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java [hunk 21]
```diff
@@ -0,0 +1,365 @@
...
+   */
+  private static List<String> getColumnNames(List<ETDataExtensionColumn> columns) {
+    List<String> columnNames = new ArrayList<String>();
+    for (ETDataExtensionColumn column : columns) {
+      columnNames.add(column.getName());
+    }
+    return columnNames;
+  }
+
+  @Override
```
Comment (avirmani) [inline src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:415]: are these empty implementations? Returning null is ok? Can you add a comment, pls

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java (hunk 1 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:43]: Updated with additional details.

File: src/main/java/com/custom/fuelsdk/ETNotSentEvent.java (hunk 2 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/ETNotSentEvent.java:33]: This is one of the tracking events that's supported by the SFMC SOAP API but was not implemented in fuelSDK java api.  The fix contains the implementation for this event type.

File: src/main/java/com/custom/fuelsdk/ETClickEvent.java (hunk 3 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/ETClickEvent.java:63]: The original ETClickEvent from FuelSDK was copied to fix a bug with methods getURLID.   These comments were part of the original code.

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java (hunk 4 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:83]: No. Usually you would specify either one of them.

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java (hunk 5 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:155]: All this code was a copy from the original ETSoapObject and extended to fix the bug.

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java (hunk 6 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java:86]: As indicated above the code is a copy from ETSoapObject and just the customRetrieve method was fixed for the bug

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java (hunk 7 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java]: Added

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java (hunk 12 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java]: Done

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java (hunk 14 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java]: Done

File: src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java (hunk 15 above)
Comment (shiva-cirus) [reply src/main/java/com/custom/fuelsdk/PaginationETSoapObject.java]: Removed the comment. This is the fix(temp) till FuelSDK team will provides a fix in an Offical release.

**Review by rmstar — APPROVED** [review]

**Pull request merged**